### PR TITLE
Several changes to improve Consensus stability:

### PR DESCRIPTION
**Pull request by mtrippled**
* Verify accepted ledger becomes validated, and retry with a new consensus transaction set if not.
 * Always store proposals.
 * Track proposals by ledger sequence. This helps slow peers catch up with the rest of the network.
 * Acquire transaction sets for proposals with future ledger sequences. This also helps slow peers catch up.
 * Optimize timer delay for establish phase to wait based on how long validators have been sending proposals. This also helps slow peers to catch up.
* Fix impasse achieving close time consensus.
* Don't wait between open and establish phases.

This PR is related to 2 others that, when combined, increase throughput significantly:
https://github.com/XRPLF/rippled/pull/4503
https://github.com/XRPLF/rippled/pull/4504

## High Level Overview of Change
There are several changes to improve consensus stability--namely, helping nodes to stay in sync as transaction volume increases. Analysis under high load shows that rippled often falls out of sync and is unable to get back into sync with the rest of the network. Namely, once rippled falls out of sync, it then attempts to acquire SHAMaps for the subsequent ledgers. As volume increases, this becomes more difficult. Nodes can fall out of sync for a variety of reasons--often having to do with intermittent slowdowns that occur as volume increases, such as garbage collection, network processing delays, and various others. Instead of fixing those issues (which should be addressed separately), this patch focuses on making rippled more robust in the event of intermittent delays. This patch has several fixes that keep a node in sync even if it falls slightly behind the rest of the network.

From the commit message:
 * Verify accepted ledger becomes validated, and retry with a new consensus transaction set if not.
 * Always store proposals.
 * Track proposals by ledger sequence. This helps slow peers catch up with the rest of the network.
 * Acquire transaction sets for proposals with future ledger sequences. This also helps slow peers catch up.
 * Optimize timer delay for establish phase to wait based on how long validators have been sending proposals. This also helps slow peers to catch up.
* Fix impasse achieving close time consensus.
Don't wait between open and establish phases.

### Context of Change
* Verify accepted ledger becomes validated, and retry with a new consensus transaction set if not:

Analysis under high volume shows that individual peers sometimes come to consensus on a position that the rest of the network discards in favor of another. This causes rippled to build a ledger that never becomes validated and falls out of sync. This could be caused by several peers changing their position but which is not received by another peer until after it comes to consensus on the wrong position. The fix for this is to essentially check if the newly built ledger has become the validated ledger. If not, look for a newer position to possibly adopt, and then retry. This step often introduces a delay in the accept phase. To counteract this, the patch measures the time delayed this way and then factors it into the open phase interval timing.

Essentially, this is implemented by splitting the the current doAccept() function into 2 functions:
buildAndValidate() and prepareOpenLedger(). Once buildAndValidate() produces a confirmed validated ledger, prepareOpenLedger() finishes up the accept phase and then launches a new consensus round. This relies upon calling updateOurPositions() in case we've built something that needs to be discarded. Care is taken to only send a single validation, and to send no new proposals after the initial decision to build and validate a ledger. This is because sending conflicting messages like this to the rest of the network could cause problems with consensus as a whole. However, it's OK to abandon a position in favor of something that actually achieves network consensus.

To illustrate, take a network with 5 validators: A,B,C,D,E. A sees proposal z from B,C,D that it agrees with and then builds a ledger. However, B,C,D changed their minds to agree with E on proposal y. With the fix, A sees that z never becomes validated and discovers that y instead is more popular. It retries with y and stays in sync with the network.

This effort has been made complex by the abstraction in the code distinguishing Consensus in an abstract sense from the XRP Ledger implementation (called RCLConsensus). This abstraction is intended allow other products to be built using the same consensus mechanism but with such things as different transaction formats. There only implementation like this is a test implementation within the rippled codebase. Note that this is distinct from other unit tests that test consensus. There is no real world use of this abstraction. This adds complexity because these 2 classes that have to interact very closely with each other. The main design implication of this has to do with the current doAccept() existing as part of RCLConsensus. However, the new implementation requires buildAndValidate()'s parameters to possibly change based on the result of updateOurPositions(). This essentially means that buildAndValidate() is now in Consensus while the remainder (prepareOpenLedger()) stays in RCLConsensus. However, all existing tests and functionality still work, including the abstraction.

 * Always store proposals.
For example, rippled currently doesn't consider proposals received during the accept phase. But this is necessary for changing the ledger that we need to work on. On a similar note, transaction sets are also discarded during the accept phase.

* Fix impasse achieving close time consensus.
A corner case was discovered in the existing code where close time consensus never is reached. The fix is described in updateOurPositions() in Consensus.h.

The remaining fixes are related to the same thing. Namely, rippled currently doesn't track peer proposals robustly enough to stay in sync after a slight delay:

 * Track proposals by ledger sequence. This helps slow peers catch up with the rest of the network.
Currently, proposals are stored in a map by peer nodeid -> deque<position>. We know the ledger upon which the proposal is based by its hash, but not necessarily by ledger sequence. The fix implements an optional new field in proposal messages to assign the ledger sequence. This allows us to know exactly which proposals correspond to ledger sequence.

 * Acquire transaction sets for proposals with future ledger sequences. This also helps slow peers catch up.
Related to this is that we now acquire transaction sets for each proposal.

After an intermittent slowdown, we will already have what are now current proposals and transaction sets.

 * Optimize timer delay for establish phase to wait based on how long validators have been sending proposals. This also helps slow peers to catch up.
The establish phase takes a minimum of 1950ms. This is to allow the whole network to spend plenty of time reconciling transaction sets.. However, if peers have already been sending proposals for at least 1950ms, then it means that we are probably lagging behind. Therefore, we should try to find consensus right now. This is implemented by placing a timestamp on proposals as received. Then, a heuristic in phaseEstablish() determines the start of the countdown to 1950ms based on proposals received vs when we first adopted a position. There is protection against byzantine validators. The implementation comments provide precise explanation.

* Don't wait between open and establish phases.
There's a 1s pause for the heartbeat timer once the ledger is closed. This is no good for a peer that is falling behind.

### Type of Change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by mtrippled** [comment]
@HowardHinnant @ChronusZ I just pushed a change and rebased it -- I hadn't seen that you guys were reviewiing it yet before pushing. If you haven't started looking at the PR, nevermind. If you have started looking at it, reload the latest. I won't make any more changes except as part of the review. What I added was something I though I had added before, but didn't, and I caught it in testing today. It doesn't change the overall scope of the project.

File: src/ripple/app/consensus/RCLConsensus.cpp
Comment (HowardHinnant) [inline src/ripple/app/consensus/RCLConsensus.cpp:387]:
```
src/ripple/app/consensus/RCLConsensus.cpp:387:27: warning: expression result unused [-Wunused-value]
    initialLedger->info().seq;
    ~~~~~~~~~~~~~~~~~~~~~ ^~~
```

File: src/ripple/app/consensus/RCLConsensus.h
Comment (HowardHinnant) [inline src/ripple/app/consensus/RCLConsensus.h:63]: If I return these data members to private, this commit still compiles for me.

File: src/ripple/app/consensus/RCLConsensus.h
Comment (HowardHinnant) [inline src/ripple/app/consensus/RCLConsensus.h:105]: If I return `validating_` back to private, this commit still compiles for me.

File: src/ripple/app/consensus/RCLConsensus.cpp
Comment (HowardHinnant) [inline src/ripple/app/consensus/RCLConsensus.cpp:921]: Avoid escaping the chrono type system.  Safer and simpler:
```c++
        (!start || std::chrono::steady_clock::now() - *start < std::chrono::seconds{5});
```

File: src/ripple/consensus/Consensus.h
Comment (HowardHinnant) [inline src/ripple/consensus/Consensus.h:745]: Just checking:  30min?  Not 3min?

**Review by HowardHinnant — CHANGES_REQUESTED** [review]
Left some comments that I think need to be addressed before approval.  But I didn't see any showstoppers.  Looks good!

File: src/ripple/app/consensus/RCLConsensus.cpp
Comment (mtrippled) [reply src/ripple/app/consensus/RCLConsensus.cpp]: fixed

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:747]: 30. I left a comment.

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h]: fixed

File: src/ripple/app/consensus/RCLConsensus.h
Comment (mtrippled) [reply src/ripple/app/consensus/RCLConsensus.h:98]: I left a comment describing why.

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:908]: fixed

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h]: fixed

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:1777]: I left an explanation in the code about the change, from lines 1740-8.

File: src/ripple/consensus/ConsensusParms.h
Comment (mtrippled) [reply src/ripple/consensus/ConsensusParms.h:76]: This logic has always been a floating point. The change I made calculates it into a variable so that it can be re-used in the code. Anyway, I've never heard of std::ratio until right now and prefer not to re-implement perfectly functioning consensus code as my first use of it.

File: src/ripple/app/consensus/RCLConsensus.h
Comment (mtrippled) [reply src/ripple/app/consensus/RCLConsensus.h]: fixed

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h]: I left a comment clarifying.

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:1455]: To protect from faulty peers on the UNL, discard the earliest proposals beyond the quorum threshold. For example, with a UNL of 20 and quorum of 80%, we should ignore the first 4 proposals received for this calculation.

**Comment by HowardHinnant** [comment]
I'm getting a repeatable unit test error:
```
ripple.conditions.PreimageSha256 Known Vectors
Assertion failed: (id != result_->position.position()), function gotTxSet, file Consensus.h, line 992.
Abort trap: 6
```

I only get the error when I run all of the unit tests:
```
rippled --unittest
```

I don't get it when I run the test by itself:
```
rippled --unittest=conditions
```

There's probably a global setting that a previous test is setting and not resetting back.

**Comment by HowardHinnant** [comment]
I've confirmed that the assert is introduced with bb74967af263fcf041d185e3094479d430092607.  It doesn't occur prior to this commit and does afterwards.  I didn't detect it previously because I wasn't running with assert enabled.

**Comment by mtrippled** [comment]
> I've confirmed that the assert is introduced with [bb74967](https://github.com/XRPLF/rippled/commit/bb74967af263fcf041d185e3094479d430092607). It doesn't occur prior to this commit and does afterwards. I didn't detect it previously because I wasn't running with assert enabled.

I confirm this. Funny, the stack trace is in Consensus_test.cpp, which comes right after the "conditions" test. It makes sense for this assertion to come as a result of Consensus. I'm not sure why the crash happened during conditions, but anyway, I'll figure it out. Thanks for finding it. Going forward, I'll do debug builds locally to find this in the future. Recently, I had been doing release builds locally once we moved to Conan, and it didn't occur to me to make the build debug.

**Comment by HowardHinnant** [comment]
I just recently learned to add this:  ` -Dassert=ON` after my `-DCMAKE_BUILD_TYPE=Release`, so I get release speed but with assert checking.

**Comment by mtrippled** [comment]
> I'm getting a repeatable unit test error:
> 
> ```
> ripple.conditions.PreimageSha256 Known Vectors
> Assertion failed: (id != result_->position.position()), function gotTxSet, file Consensus.h, line 992.
> Abort trap: 6
> ```
> 
> I only get the error when I run all of the unit tests:
> 
> ```
> rippled --unittest
> ```
> 
> I don't get it when I run the test by itself:
> 
> ```
> rippled --unittest=conditions
> ```
> 
> There's probably a global setting that a previous test is setting and not resetting back.

Fixed, and all unittests pass with no assertions.

**Review by HowardHinnant — APPROVED** [review]

**Comment by intelliot** [comment]
@ChronusZ will you be able to review this?

File: src/ripple/consensus/Consensus.h
Comment (ximinez) [inline src/ripple/consensus/Consensus.h:1590]: `result_` is not always set when I run the `Consensus` unit test. It asserts in Windows. Since `phaseEstablish` starts with `assert(result_);`, I can only assume it's getting prematurely cleared somewhere.

Here's the stack
```
 	rippled.exe!std::optional<ripple::ConsensusResult<ripple::test::csf::Peer>>::operator*() Line 359	C++
>	rippled.exe!ripple::Consensus<ripple::test::csf::Peer>::phaseEstablish() Line 1587	C++
 	rippled.exe!ripple::Consensus<ripple::test::csf::Peer>::timerEntry(const std::chrono::time_point<ripple::NetClock,std::chrono::duration<unsigned int,std::ratio<1,1>>> & now) Line 986	C++
 	rippled.exe!ripple::test::csf::Peer::timerEntry() Line 932	C++
 	rippled.exe!`ripple::test::csf::Peer::timerEntry'::`4'::<lambda_1>::operator()() Line 933	C++
 	rippled.exe!ripple::test::csf::Scheduler::event_impl<`ripple::test::csf::Peer::timerEntry'::`4'::<lambda_1>>::operator()() Line 106	C++
 	rippled.exe!ripple::test::csf::Scheduler::step_one() Line 398	C++
 	rippled.exe!ripple::test::csf::Scheduler::step() Line 408	C++
 	rippled.exe!ripple::test::Consensus_test::testStandalone() Line 135	C++
 	rippled.exe!ripple::test::Consensus_test::run() Line 1058	C++
 	rippled.exe!beast::unit_test::suite::run<void>(beast::unit_test::runner & r) Line 589	C++
 	rippled.exe!beast::unit_test::suite::operator()<void>(beast::unit_test::runner & r) Line 429	C++
 	rippled.exe!`beast::unit_test::make_suite_info<ripple::test::Consensus_test>'::`2'::<lambda_1>::operator()(beast::unit_test::runner & r) Line 120	C++
 	[External Code]	
 	rippled.exe!beast::unit_test::suite_info::run(beast::unit_test::runner & r) Line 87	C++
 	rippled.exe!beast::unit_test::runner::run<void>(const beast::unit_test::suite_info & s) Line 188	C++
 	rippled.exe!ripple::test::multi_runner_child::run_multi<ripple::multi_selector>(ripple::multi_selector pred) Line 337	C++
 	rippled.exe!ripple::runUnitTests(const std::string & pattern, const std::string & argument, bool quiet, bool log, bool child, bool ipv6, unsigned __int64 num_jobs, int argc, char * * argv) Line 282	C++
 	rippled.exe!ripple::run(int argc, char * * argv) Line 547	C++
 	rippled.exe!main(int argc, char * * argv) Line 813	C++
 	[External Code]	
```

**Review by ximinez — CHANGES_REQUESTED** [review]
If you remove the `#include`s that I flagged, it resolves any spurious circular dependencies that were introduced, but you'll still need to run `./Builds/levelization/levelization.sh` and commit the changes to `ordering.txt`, since new dependencies were introduced to make the levelization check job happy.

**Comment by intelliot** [comment]
note: @mtrippled is currently working on this

File: src/ripple/proto/ripple.proto
Comment (mtrippled) [reply src/ripple/proto/ripple.proto]: fixed

**Comment by mtrippled** [comment]
> If you remove the `#include`s that I flagged, it resolves any spurious circular dependencies that were introduced, but you'll still need to run `./Builds/levelization/levelization.sh` and commit the changes to `ordering.txt`, since new dependencies were introduced to make the levelization check job happy.

fixed

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:899]: I tried to a little bit, but ended up in a rat hole pretty quickly.

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:1471]: It's correct as it is. UNL of 20 with 80% quorum is 16. Discard the remainder.
4 = (16 / 0.8) - 16

I put more detail in the comment.

File: src/ripple/overlay/impl/PeerImp.cpp
Comment (mtrippled) [reply src/ripple/overlay/impl/PeerImp.cpp]: fixed

File: src/ripple/overlay/impl/PeerImp.cpp
Comment (mtrippled) [reply src/ripple/overlay/impl/PeerImp.cpp:43]: LedgerIndex is in Protocol.h.

File: src/ripple/overlay/impl/PeerImp.cpp
Comment (mtrippled) [reply src/ripple/overlay/impl/PeerImp.cpp]: fixed

File: src/test/csf/Peer.h
Comment (mtrippled) [reply src/test/csf/Peer.h]: fixed

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:1639]: Weird. I put a commit here that prints a bunch of things to stderr that I hope sheds light. Can you please try again? The commit is 4e119da886824c4f91f3512929428b98bf554c7c.

The only thing I can think of is that the unittest calls timerEntry outside of a lock, so there's a possible race condition between setting the phase to establish and setting the result_ in closeLedger(). It's possible I introduced something that triggers the race condition from the standpoint of the test. I know that I'm calling timerEntry() directly now to save time during consensus. The stderr output I added should confirm this is happening.

**Review by ximinez — CHANGES_REQUESTED** [review]
Debug output on Windows:
```
$ ./build/cmake/msvc19.OFF/Debug/rippled --unittest=Consensus

Peer.h timerEntry calling timerEntry
timerEntry phase open phase unchanged after checkLedger
closeLedger transitioned to establish
closeLedger set result_
closeLedger calling timerEntry
timerEntry phase establish phase unchanged after checkLedger
Peer.h timerEntry scheduler calling timerEntry
Peer.h timerEntry calling timerEntry
timerEntry phase establish phase unchanged after checkLedger
Peer.h timerEntry scheduler calling timerEntry
Peer.h timerEntry calling timerEntry
timerEntry phase establish phase unchanged after checkLedger
ripple.consensus.Consensus
```
Additionally, try putting `assert(result)` just before the call to `adaptor_.onAccept` (Consensus.h, line 1634). I vaguely remember something like this happening before, where the Windows implementation asserts if the optional is not seated, while other OSs don't, even though it's not seated there either. If that's the case, it may help you debug faster.

**Comment by mtrippled** [comment]
> Debug output on Windows:
> 
> ```
> $ ./build/cmake/msvc19.OFF/Debug/rippled --unittest=Consensus
> 
> Peer.h timerEntry calling timerEntry
> timerEntry phase open phase unchanged after checkLedger
> closeLedger transitioned to establish
> closeLedger set result_
> closeLedger calling timerEntry
> timerEntry phase establish phase unchanged after checkLedger
> Peer.h timerEntry scheduler calling timerEntry
> Peer.h timerEntry calling timerEntry
> timerEntry phase establish phase unchanged after checkLedger
> Peer.h timerEntry scheduler calling timerEntry
> Peer.h timerEntry calling timerEntry
> timerEntry phase establish phase unchanged after checkLedger
> ripple.consensus.Consensus
> ```
> 
> Additionally, try putting `assert(result)` just before the call to `adaptor_.onAccept` (Consensus.h, line 1634). I vaguely remember something like this happening before, where the Windows implementation asserts if the optional is not seated, while other OSs don't, even though it's not seated there either. If that's the case, it may help you debug faster.

I believe this is now fixed. I squashed, rebased, and force-pushed the recent debugging commits and removed the std::cerr output. Please try again.

**Review by ximinez — APPROVED** [review]
I left a couple of comments, but nothing that will stop me from approving this either way.

You do need to merge your branch with `develop` so that CI can run.

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h]: fixed

File: src/ripple/consensus/Consensus.h
Comment (mtrippled) [reply src/ripple/consensus/Consensus.h:1575]: fixed

**Comment by mtrippled** [comment]
> 

Yes, I need to clang-format, too. I'll do that.

**Comment by mtrippled** [comment]
@HowardHinnant can you please review the changes I've made since your previous review? Ed found some things.

**Review by HowardHinnant — APPROVED** [review]

**Comment by intelliot** [comment]
(note: @manojsdoshi is working with Mark to test and merge this PR along with #4503 and #4504)

**Comment by manojsdoshi** [comment]
@ximinez @HowardHinnant can one of you guys help resolve this merge conflict?

**Comment by ximinez** [comment]
> can one of you guys help resolve this merge conflict?

@intelliot Conflict was pretty trivial. I verified it builds, but I'll let CI handle the tests.

**Comment by intelliot** [comment]
Since this was reverted in 1.12.0-rc3, this PR is being reopened for additional review, testing, and potential improvements, with the expectation that it will be merged again soon (e.g. for the 1.13 or 2.0 release).

**Comment by intelliot** [comment]
in progress: considering a txThreshold so new behavior is used under high volume. change would likely be pushed to this `consensus` branch/PR.

**Pull request merged**
